cmd/openstack-cloud-controller-manager/main.go

```diff
@@ -23,115 +23,46 @@ import (
 	"k8s.io/klog/v2"
 	_ "k8s.io/kubernetes/pkg/features" // add the kubernetes feature gates
 
-	"github.com/spf13/cobra"
-	"github.com/spf13/pflag"
-)
-
-func init() {
-	mux := http.NewServeMux()
-	healthz.InstallHandler(mux)
-}
-
-var (
-	versionFlag bool
+	"k8s.io/cloud-provider-openstack/pkg/openstack"
+	"k8s.io/cloud-provider-openstack/pkg/version"
 )
 
 func main() {
-	rand.Seed(time.Now().UTC().UnixNano())
-
-	goflag.CommandLine.Parse([]string{})
-	controllerList := []string{"cloud-node", "cloud-node-lifecycle", "service", "route"}
+	rand.Seed(time.Now().UnixNano())
 
-	s, err := options.NewCloudControllerManagerOptions()
+	ccmOptions, err := options.NewCloudControllerManagerOptions()
 	if err != nil {
 		klog.Fatalf("unable to initialize command options: %v", err)
 	}
-	s.KubeCloudShared.CloudProvider.Name = openstack.ProviderName
-
-	command := &cobra.Command{
-		Use: "openstack-cloud-controller-manager",
-		Long: `The Cloud controller manager is a daemon that embeds
-the cloud specific control loops shipped with Kubernetes.`,
-		PersistentPreRun: func(cmd *cobra.Command, args []string) {
-			// Glog requires this otherwise it complains.
-			goflag.CommandLine.Parse(nil)
-
-			// This is a temporary hack to enable proper logging until upstream dependencies
-			// are migrated to fully utilize klog instead of glog.
-			klogFlags := goflag.NewFlagSet("klog", goflag.ExitOnError)
-			klog.InitFlags(klogFlags)
-
-			// Sync the glog and klog flags.
-			cmd.Flags().VisitAll(func(f1 *pflag.Flag) {
-				f2 := klogFlags.Lookup(f1.Name)
-				if f2 != nil {
-					value := f1.Value.String()
-					f2.Value.Set(value)
-				}
-			})
-		},
-		Run: func(cmd *cobra.Command, args []string) {
-			if versionFlag {
-				version.PrintVersionAndExit()
-			}
-
-			flag.PrintFlags(cmd.Flags())
-
-			c, err := s.Config(controllerList, app.ControllersDisabledByDefault.List())
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
-				os.Exit(1)
-			}
-
-			cloudconfig := c.Complete().ComponentConfig.KubeCloudShared.CloudProvider
-			cloud, err := cloudprovider.InitCloudProvider(cloudconfig.Name, cloudconfig.CloudConfigFile)
-			if err != nil {
-				klog.Fatalf("Cloud provider could not be initialized: %v", err)
-			}
-			if cloud == nil {
-				klog.Fatalf("cloud provider is nil")
-			}
-
-			if err := app.Run(c.Complete(), app.DefaultControllerInitializers(c.Complete(), cloud), wait.NeverStop); err != nil {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
-				os.Exit(1)
-			}
-		},
-	}
 
-	fs := command.Flags()
-	namedFlagSets := s.Flags(controllerList, app.ControllersDisabledByDefault.List())
-	for _, f := range namedFlagSets.FlagSets {
-		fs.AddFlagSet(f)
-	}
-
-	fs.BoolVar(&versionFlag, "version", false, "Print version and exit")
+	fss := cliflag.NamedFlagSets{}
+	command := app.NewCloudControllerManagerCommand(ccmOptions, cloudInitializer, app.DefaultInitFuncConstructors, fss, wait.NeverStop)
```

Comment:
is this going to change cobra.command from `openstack-cloud-controller-manager` to `cloud-controller-manager`? if yes, we should perhaps modify manifests also?

Reply:
Nope, the command is not changed, but the help message will look weird as it's saying:

```
Usage:
  cloud-controller-manager [flags]
```

Reply:
@ramineni I don't think we could resolved the `--version` issue perfectly as it's defined in `NewCloudControllerManagerCommand` method. 

```
$ ./openstack-cloud-controller-manager --version
Kubernetes v0.0.0-master+$Format:%H$
```

One workaround I can think of is to define a different param (e.g. `occm-version`) to maintain our own version, or we leave it to k8s community to provide a mechanism for cloud providers. What do you think?

cc @jichenjc

Reply:
> @ramineni I don't think we could resolved the --version issue perfectly as it's defined in NewCloudControllerManagerCommand method.

@lingxiankong right. From here it looked to me that it should take the values from ldflags , https://github.com/kubernetes/component-base/blob/master/version/version.go#L29 , but doesnt seem to work .

> One workaround I can think of is to define a different param (e.g. occm-version) to maintain our own version, or we leave it to k8s community to provide a mechanism for cloud providers. What do you think?

@lingxiankong +1, I checked a lil bit on this before , couldnt get it working as not clean mechanism provided yet . Ill get to this option again . Let me also check with cloudprovider team if they have any other options. Thanks.
I'll raise a bug to track this issue in meantime

Reply:
>>>One workaround I can think of is to define a different param (e.g. occm-version) to maintain our own version, or we leave it to k8s community to provide a mechanism for cloud providers. What do you think?

+1 to me as well, we can provide some quick solution for now and overtime let's switch to cloud-provider way